### DISCUSSION TITLE
ASUS H87I-PLUS Can't flash modified BIOS; DSDT Patching not working.

### DISCUSSION THREAD

**Discussion by englut** [xCuri0/ReBarUEFI#133] [General]
**System**
 - Motherboard: ASUS H87I-PLUS
 - BIOS Version: 2003 (2014-11-05)
 - GPU:
 - [x] CSM is turned off. **Make sure to confirm this in the BIOS and not with GPU-Z or similar since it can be inaccurate**
 - [x] 4G decoding is enabled. **Make sure to confirm this in the BIOS and not with GPU-Z or similar since it can be inaccurate**
 - [x] UEFIPatch is applied (see [Using UEFIPatch](https://github.com/xCuri0/ReBarUEFI/wiki/Using-UEFIPatch) for more information). On some motherboards [DSDT Patching](https://github.com/xCuri0/ReBarUEFI/wiki/DSDT-Patching) is also needed
 - [x] I have read [Common issues (and fixes)](https://github.com/xCuri0/ReBarUEFI/wiki/Common-issues-(and-fixes))

**Description**
My ASUS motherboard doesn't support BIOS Flashback, so I attempted to flash via the BIOS interface (ASUS EZ Flash 2 Utility v02.00). When I try to flash the modified BIOS however, I get a `Security verification failed.` error.

I'm not sure how to bypass this security feature. I've tried another method using the ASUS DOS bupdater.exe, but I get the same issue, and I can't even do a BIOS backup. I have Secure Boot set to `Other OS`, there's no Disabled state from what I can tell. 

Does this mean I'm SOL on getting Resizable BAR enabled on my motherboard?

Specifications: https://dlcdnets.asus.com/pub/ASUS/mb/LGA1150/H87I-PLUS/E8210_H87I-PLUS.pdf

4G decoding doesn't seem to be enabled or available on my motherboard. I tried to find the `4G Decod` or `Above 4G` but only seem to find Memory Remapping.

This is the output of `lspci -v` of my system:
```
03:00.0 VGA compatible controller: Intel Corporation Device 56a6 (rev 05) (prog-if 00 [VGA controller])
	Subsystem: ASRock Incorporation Device 6007
	Flags: bus master, fast devsel, latency 0, IRQ 28
	Memory at f6000000 (64-bit, non-prefetchable) [size=16M]
	Memory at e0000000 (64-bit, prefetchable) [size=256M]
	Expansion ROM at f7000000 [disabled] [size=2M]
	Capabilities: [40] Vendor Specific Information: Len=0c <?>
	Capabilities: [70] Express Endpoint, MSI 00
	Capabilities: [ac] MSI: Enable+ Count=1/1 Maskable+ 64bit+
	Capabilities: [d0] Power Management version 3
	Capabilities: [100] Alternative Routing-ID Interpretation (ARI)
	Capabilities: [420] Physical Resizable BAR
	Capabilities: [400] Latency Tolerance Reporting
	Kernel driver in use: i915
	Kernel modules: i915
```

This is the output of `dmesg | grep "root bus resource"`:
```
[    0.217132] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    0.217135] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    0.217137] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000dffff window]
[    0.217138] pci_bus 0000:00: root bus resource [mem 0xe0000000-0xfeafffff window]
[    0.217140] pci_bus 0000:00: root bus resource [bus 00-3e]
```

-----

**UPDATE:** Attempted the [DSDT patching](https://github.com/xCuri0/ReBarUEFI/wiki/DSDT-Patching#dsdt-broken-4g-decoding-fix-haswell-and-up--linux-no-4g-decoding-fix) on Linux. I've got the DSDT patch running, but `dmesg` is still showing that Resizable BAR isn't enabled and I'm stuck at 256MB.

```
$ dmesg | grep BAR
[    2.043241] i915 0000:03:00.0: BAR 0: releasing [mem 0xf6000000-0xf6ffffff 64bit]
[    2.043245] i915 0000:03:00.0: BAR 2: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.043266] pcieport 0000:02:01.0: BAR 15: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.043270] pcieport 0000:01:00.0: BAR 15: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.043277] pcieport 0000:01:00.0: BAR 15: no space for [mem size 0x100000000 64bit pref]
[    2.043280] pcieport 0000:01:00.0: BAR 15: failed to assign [mem size 0x100000000 64bit pref]
[    2.043283] pcieport 0000:02:01.0: BAR 15: no space for [mem size 0x100000000 64bit pref]
[    2.043296] pcieport 0000:02:01.0: BAR 15: failed to assign [mem size 0x100000000 64bit pref]
[    2.043299] i915 0000:03:00.0: BAR 2: no space for [mem size 0x100000000 64bit pref]
[    2.043301] i915 0000:03:00.0: BAR 2: failed to assign [mem size 0x100000000 64bit pref]
[    2.043305] i915 0000:03:00.0: BAR 0: assigned [mem 0xf6000000-0xf6ffffff 64bit]
[    2.043396] i915 0000:03:00.0: [drm] Failed to resize BAR2 to 4096M (-ENOSPC)
[    2.043401] i915 0000:03:00.0: BAR 2: assigned [mem 0xe0000000-0xefffffff 64bit pref]
[    2.043426] i915 0000:03:00.0: [drm] Using a reduced BAR size of 256MiB. Consider enabling 'Resizable BAR' or similar, if available in the BIOS.
```

Here's the latest output of `dmesg | grep "root bus resource"`:
```
$ dmesg | grep "root bus resource"
[    0.176629] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    0.176632] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    0.176635] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000dffff window]
[    0.176638] pci_bus 0000:00: root bus resource [mem 0xe0000000-0xfeafffff window]
[    0.176641] pci_bus 0000:00: root bus resource [mem 0x41f000000-0x7fffffffff window]
[    0.176644] pci_bus 0000:00: root bus resource [bus 00-3e]
```

I've also attempted to use `ReBarState` and have set the value to 11 / 2048 (not trying to be greedy 🥲):
```
$ sudo ./ReBarState
ReBarState (c) 2022 xCuri0

Current ReBarState 11 / 2048 MB
```

**Comment by Daishuku** [comment]
Some other guy told me, ASUS BIOS.rom has their own header that design for EZ Flash, need convert to .cap to remove header, maybe the the main factor cause UEFI tools ruin ASUS motherboard.

**Comment by englut** [comment]
> Some other guy told me, ASUS BIOS.rom has their own header that design for EZ Flash, need convert to .cap to remove header, maybe the the main factor cause UEFI tools ruin ASUS motherboard.

@Daishuku Ah so that means I'm probably SOL then?

**Comment by xCuri0** [comment]
There is no reason for you to flash BIOS if you use Linux, all you have to do is the Linux DSDT fix where you add a QWordMemory entry which the wiki explains

**Reply by englut** [reply]
You're referring to this section?
https://github.com/xCuri0/ReBarUEFI/wiki/DSDT-Patching#dsdt-broken-4g-decoding-fix-haswell-and-up--linux-no-4g-decoding-fix
https://wiki.archlinux.org/title/DSDT#Using_modified_code

I'll give it a shot! Thank you!

**Reply by englut** [reply]
@xCuri0 I'm going through that section again, and I'm unclear what this portion means:

```You can even override DSDT without any BIOS modification on Linux, just make sure to use DSDT from /sys/firmware/acpi/tables/DSDT as a base if you're using this method.```

Specifically this `just make sure to use DSDT from /sys/firmware/acpi/tables/DSDT as a base if you're using this method.`.

Could you elaborate? Do you mean to use this as the DSDT.aml? Rather than the one extracted from the modified bios up to this point in the wiki?

**Comment by englut** [comment]
@xCuri0 I've completed the patched ACPI boot method! But I'm still not getting resizable bar. 

I've confirmed that my kernel is running the overridden DSDT patch. 

Before patch:
```
$ sudo dmesg | grep DSDT
[    0.008404] ACPI: DSDT 0x00000000DD379188 00DDA1 (v02 ALASKA A M I    00000031 INTL 20091112)
[    0.008439] ACPI: Reserving DSDT table memory at [mem 0xdd379188-0xdd386f28]
```

After patch:
```
$ sudo dmesg | grep DSDT
[    0.009158] ACPI: DSDT 0x00000000DD366000 00E05E (v02 ALASKA A M I    00000032 INTL 20200925)
[    0.009175] ACPI: Reserving DSDT table memory at [mem 0xdd366000-0xdd37405d]
```


But it doesn't seem like Resizable BAR is enabled (even after using `ReBarState`).
```
$ sudo dmesg | grep BAR
[    0.220151] pci 0000:03:00.0: BAR 2: assigned to efifb
[    2.141309] i915 0000:03:00.0: BAR 0: releasing [mem 0xf6000000-0xf6ffffff 64bit]
[    2.141313] i915 0000:03:00.0: BAR 2: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.141336] pcieport 0000:02:01.0: BAR 15: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.141340] pcieport 0000:01:00.0: BAR 15: releasing [mem 0xe0000000-0xefffffff 64bit pref]
[    2.141349] pcieport 0000:01:00.0: BAR 15: no space for [mem size 0x100000000 64bit pref]
[    2.141352] pcieport 0000:01:00.0: BAR 15: failed to assign [mem size 0x100000000 64bit pref]
[    2.141356] pcieport 0000:02:01.0: BAR 15: no space for [mem size 0x100000000 64bit pref]
[    2.141360] pcieport 0000:02:01.0: BAR 15: failed to assign [mem size 0x100000000 64bit pref]
[    2.141363] i915 0000:03:00.0: BAR 2: no space for [mem size 0x100000000 64bit pref]
[    2.141366] i915 0000:03:00.0: BAR 2: failed to assign [mem size 0x100000000 64bit pref]
[    2.141370] i915 0000:03:00.0: BAR 0: assigned [mem 0xf6000000-0xf6ffffff 64bit]
[    2.141439] i915 0000:03:00.0: [drm] Failed to resize BAR2 to 4096M (-ENOSPC)
[    2.141444] i915 0000:03:00.0: BAR 2: assigned [mem 0xe0000000-0xefffffff 64bit pref]
[    2.141475] i915 0000:03:00.0: [drm] Using a reduced BAR size of 256MiB. Consider enabling 'Resizable BAR' or similar, if available in the BIOS.
```

```
$ sudo lspci -v | grep -i 'memory at'
	Memory at f7420000 (64-bit, non-prefetchable) [size=64K]
	Memory at f743b000 (64-bit, non-prefetchable) [size=16]
	Memory at f7400000 (32-bit, non-prefetchable) [size=128K]
	Memory at f7439000 (32-bit, non-prefetchable) [size=4K]
	Memory at f7438000 (32-bit, non-prefetchable) [size=1K]
	Memory at f7430000 (64-bit, non-prefetchable) [size=16K]
	Memory at f7437000 (32-bit, non-prefetchable) [size=1K]
	Memory at f7436000 (32-bit, non-prefetchable) [size=2K]
	Memory at f7435000 (64-bit, non-prefetchable) [size=256]
	Memory at f0000000 (64-bit, prefetchable) [size=8M]
	Memory at f6000000 (64-bit, non-prefetchable) [size=16M]
	Memory at e0000000 (64-bit, prefetchable) [size=256M]
	Memory at f7300000 (64-bit, non-prefetchable) [size=16K]
```

Any idea what I may be doing incorrectly?

**Reply by englut** [reply]
Forgot to include this output:
```
$ dmesg | grep "root bus resource"
[    0.217478] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    0.217481] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    0.217485] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000dffff window]
[    0.217488] pci_bus 0000:00: root bus resource [mem 0xe0000000-0xfeafffff window]
[    0.217491] pci_bus 0000:00: root bus resource [mem 0x41f000000-0x7fffffffff window]
[    0.217495] pci_bus 0000:00: root bus resource [bus 00-3e]
```

**Comment by englut** [comment]
@xCuri0 do you have any idea what I may be doing incorrectly? 🙏

**Comment by histfuture** [comment]
Same question.
I am using linux, what does this mean. 

**"There is no reason for you to flash BIOS if you use Linux, all you have to do is the Linux DSDT fix where you add a QWordMemory entry which the wiki explains"**

Does this mean I do not use uefi patch or uefi tools to modify my bios if I use linux?
